Remove the -sil-serialize-all option

swiftix opened the pull request.

slavapestov committed:
Re-apply "SIL: Remove special meaning for @_semantics("stdlib_binary_only")"

With -sil-serialize-all gone, this no longer means anything; just
don't declare the function as @_inlineable instead.

Fixes <rdar://problem/34564380>.

swiftix committed:
Remove the -sil-serialize-all option

swiftix committed:
Update the tests after removing the -sil-serialize-all option

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@slavapestov Slava, do you want to have a look?

swiftix commented:
@swift-ci please test

slavapestov commented on include/swift/SIL/SILModule.h:
```diff
@@ -323,7 +323,7 @@ class SILModule {
 
   /// Returns true if everything in this SILModule is being serialized.
-  bool isWholeModuleSerialized() const { return Options.SILSerializeAll; }
+  bool isWholeModuleSerialized() const { return false; }
```
Are you going to remove this, and ResilienceStrategy::Fragile, etc?

swiftix replied:
Yes, I'm going to remove it as well. There is no need for it anymore.

slavapestov replied:
Also makeModuleFragile() in SILGen

swiftix committed:
Remove even more dead code after -sil-serialize-all is gone

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

jrose-apple commented:
StdlibUnittest was still using -sil-serialize-all in order to make sure that things actually get specialized. Are we going to do something about that? cc @moiseev

slavapestov commented:
@jrose-apple I don't think so. Can you un-revert my PR so that this one can apply cleanly?

slavapestov commented:
@swifix Once https://github.com/apple/swift/pull/12275 passes tests you can merge it and then merge this

slavapestov commented on lib/Frontend/Frontend.cpp:
```diff
@@ -258,8 +258,6 @@ ModuleDecl *CompilerInstance::getMainModule() {
     if (Invocation.getFrontendOptions().EnableResilience)
       MainModule->setResilienceStrategy(ResilienceStrategy::Resilient);
-    else if (Invocation.getSILOptions().SILSerializeAll)
-      MainModule->setResilienceStrategy(ResilienceStrategy::Fragile);
```
You forgot to remove ResiienceStrategy::Fragile

swiftix replied:
@slavapestov In cases like below, should we check instead for SILSerializeWitnessTables or remove them alltogether?  
```c++
if (isProtocolConformanceKind(getKind())) {
    auto conformance = getProtocolConformance();

    auto conformanceModule = conformance->getDeclContext()->getParentModule();
    auto isCompletelySerialized = conformanceModule->getResilienceStrategy() ==
                                  ResilienceStrategy::Fragile;

    // The conformance is fragile if it is in a -sil-serialize-all module.
    return isCompletelySerialized;
  }
```

slavapestov replied:
I think you should check your flag yes.

swiftix replied:
@slavapestov Hmm. There is a small problem here. The code above is inside `LinkEntity::isFragile`, which does not contain any reference to the `SILModule`, which has `SILSerializeWitnessTables ` flag. It can only access `ModuleDecl` through `conformance->getDeclContext()->getParentModule()`.

How do we solve this in the cleanest way? Should we add an `IGM` argument to the function? Should we add a flag for -sil-serialize-witness-tables and -sil-serialize-vtables to the `ModuleDecl` or may be even have a reference from `ModuleDecl` to `SILOptions`?

BTW, why don't `ModuleDecls` have a back-reference to the `SILModules`? It could be useful for such cases...

swiftix replied:
OK. I moved `SILSerializeWitnessTables` and `SILSerializeVTables` into `LanguageOptions`, so that they are accessible from `ASTContext` and `ModuleDecl`.

swiftix committed:
Remove ResilientStrategy::Fragile

swiftix committed:
Serialize witnesses if -sil-serialize-witness-tables is provided

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean test

swiftix commented:
@swift-ci please clean test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test Linux

swiftix commented:
@swift-ci please clean test Linux

swiftix committed:
Fix a couples of bugs related to witness method serialization

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix committed:
Fix serialization of vtables

Only vtables of public classes should be serialized. And only entries for public or serialized methods should be serialized.

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix committed:
Fix tests

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean test

swiftix commented:
@swift-ci please clean test OS X

swiftix commented:
@slavapestov Small correction. I meant: "Should we conditionally addserialized to witness thunks only when the actual implementation is public or serialized as well"?

slavapestov commented:
@swiftix Witness method thunks should only be marked as serializable if the witness is public.

slavapestov commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1338,8 +1338,8 @@ bool LinkEntity::isFragile(ForDefinition_t isDefinition) const {
-    auto isCompletelySerialized = conformanceModule->getResilienceStrategy() ==
-                                  ResilienceStrategy::Fragile;
+    auto isCompletelySerialized =
+        conformanceModule->getASTContext().LangOpts.SILSerializeWitnessTables;
```
This is wrong. LinkEntity::isFragile() should be removed, and callers that checked it should always assume 'false'.

swiftix replied:
@slavapestov Slava, LinkEntity::isFragile is currently returning true for e.g. SILFunctions that are [serialized]. Are you sure it is a good idea to always return 'false' for any inputs?

slavapestov commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -482,9 +477,10 @@ class SILGenConformance : public SILGenWitnessTable<SILGenConformance> {
-      witnessSerialized = (SGM.isMakeModuleFragile()
-                           ? IsSerialized
-                           : IsNotSerialized);
+      witnessSerialized =
```
This should be IsNotSerialized

slavapestov commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -2276,14 +2284,17 @@ void SILSerializer::writeSILBlock(const SILModule *SILMod) {
   assert(assocDC && "cannot serialize SIL without an associated DeclContext");
   for (const SILVTable &vt : SILMod->getVTables()) {
-    if ((ShouldSerializeAll || SILMod->getOptions().SILSerializeVTables) &&
+    if ((ShouldSerializeAll ||
```
I think you can remove ShouldSerializeAll also

swiftix replied:
Do you mean only in this condition or in the whole file? In spite of its name, this was not set by `-sil-serialize-all`, but by something different.

slavapestov replied:
It looks like when emitting SIB, we set it unconditionally, but otherwise it is keyed off -sil-serialize-all. I suggest removing it as well as serializationOpts.SerializeAllSIL.

slavapestov replied:
I see some other usages of this:

```
Slavas-MacBook-Pro:swift slava$ git grep SerializeAllSIL
include/swift/Serialization/SerializationOptions.h:    bool SerializeAllSIL = false;
lib/FrontendTool/FrontendTool.cpp:      serializationOpts.SerializeAllSIL = true;
lib/FrontendTool/FrontendTool.cpp:      serializationOpts.SerializeAllSIL = true;
lib/FrontendTool/FrontendTool.cpp:      serializationOpts.SerializeAllSIL =
lib/Serialization/Serialization.cpp:    S.writeSIL(SILMod, options.SerializeAllSIL);
tools/sil-func-extractor/SILFunctionExtractor.cpp:    serializationOpts.SerializeAllSIL = true;
tools/sil-opt/SILOpt.cpp:    serializationOpts.SerializeAllSIL = true;
```

I suspect they're all bogus. Please just rip it out, and since sil-func-extractor and sil-opt are not part of the usual developer flow, any tests that fail as a result are probably bogus.

swiftix commented:
@slavapestov The current logic for witness method thunks is:

```c++
    // ... or if the conformance itself thinks it should be.
    if (SILWitnessTable::conformanceIsSerialized(
            Conformance, SGM.M.getSwiftModule()->getResilienceStrategy(),
            SGM.M.getASTContext().LangOpts.SILSerializeWitnessTables))
      Serialized = IsSerialized;
```

But the mentioned test results in a crash with this logic.

slavapestov commented:
@swiftix I think the problem is in the fixmeWitnessHasLinkageThatNeedsToBePublic() case actually.

slavapestov commented:
I think this patch can be dropped: https://github.com/apple/swift/pull/12256/commits/6c85fff3896e257525eac5b9f696d530989380e2

If there are tests that rely on the thunk being serialized in this case, they should just be dropped since they're no longer valid.

slavapestov commented:
Yeah, in the dead_witness_module.swift, the witness `MyStruct.confx()` is internal, but the protocol requirement is `@_versioned`. The conformance checker should reject that. In the meantime, you should make the witness `@_versioned`.

slavapestov commented:
LinkEntity::isFragile is completely dead too

swiftix commented:
@slavapestov With all the suggested changes, we also run into another issue. Top-level functions which are emitted using `emitTopLevelFunction` are not marked as serialized. As a result, `main` is not serialized! :-) This results e.g. in a failure of `emit-sib-single-file.swift`. Should we always mark `main` (and may be all other top-level compiler-generated functions) as serialized?

jrose-apple commented:
`main` should definitely not be serialized normally, and `-emit-sib` should be preserving but not honoring the serialized bit. Remember that SIB is supposed to be an intermediate file.

swiftix committed:
Remove LinkEntity::isFragile which is always false now.

And remove all the code that is dead because of it.

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean test

swiftix commented:
@swift-ci please clean test

swiftix commented:
@swift-ci please smoke benchmark

swiftix commented:
@swift-ci please smoke benchmark

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean smoke test

swiftix commented:
@swift-ci please clean test

swiftix commented:
@swift-ci Please test source compatibility

swiftix commented:
@slavapestov The tests seem to pass finally, at least the smoke tests. OK to merge once the full tests are green? Or do we want to anything else to this PR?

swiftix commented:
@swift-ci Please test source compatibility

swiftix committed:
Output all vtables with all their entries if ShouldSerializeAll is true

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please clean test

slavapestov commented:
Looks good to go!

jrose-apple commented:
Looks like this is *still* causing problems somehow in https://ci.swift.org/job/oss-swift_tools-RA_stdlib-RD_test-simulator/180/, under optimized testing (something "please [clean] test" doesn't do). @swiftix, are you going to have time to look at this this morning or should I revert again?

jrose-apple commented:
```
Undefined symbols for architecture x86_64:
  "__T014StdlibUnittest15ResettableValueC5valuexvs", referenced from:
      __T04mainyycfU_ in main-7ebfb0.o
      __T04mainyycfU0_Tm in main-7ebfb0.o
      __T04mainyycfU2_ in main-7ebfb0.o
      __T04mainyycfU3_ in main-7ebfb0.o
      __T04mainyycfU4_ in main-7ebfb0.o
ld: symbol(s) not found for architecture x86_64
```

jrose-apple commented:
Or did you or Slava already get to it?

swiftix commented:
@jrose-apple I'll try to have a look

swiftix commented:
@jrose-apple I tried to reproduce locally using `swift/utils/build-script --preset=buildbot,tools=RA,stdlib=RD`, but I do not get this error. Something must be messed up on the bots.
